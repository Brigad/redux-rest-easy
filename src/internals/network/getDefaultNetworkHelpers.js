const DEFAULT_NETWORK_HELPERS = {
  getToken: () => 'token',
  requestGET() {
    return {
      method: 'GET',
      headers: {
        Accept: 'application/json',
        Authorization: `Bearer ${this.getToken()}`,
      },
    };
  },
  requestPATCH(body) {
    return {
      method: 'PATCH',
      headers: {
        Accept: 'application/json',
        Authorization: `Bearer ${this.getToken()}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    };
  },
  requestPUT(body) {
    return {
      method: 'PUT',
      headers: {
        Accept: 'application/json',
        Authorization: `Bearer ${this.getToken()}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    };
  },
  requestPOST(body) {
    return {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        Authorization: `Bearer ${this.getToken()}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    };
  },
  requestDELETE() {
    return {
      method: 'DELETE',
      headers: {
        Accept: 'application/json',
        Authorization: `Bearer ${this.getToken()}`,
      },
    };
  },
  handleStatusCode: (response) => {
    if (response) {
      if (response.status >= 200 && response.status < 300) {
        return response;
      }

      const error = new Error(response.statusText);
      error.response = response;
      throw error;
    }

    return null;
  },
  // eslint-disable-next-line no-unused-vars
  handleError: async (err, dispatch) => {
    try {
      if (err && err.response) {
        const error = await err.response.json();

        // dispatch some action to warn the user about the error

        // eslint-disable-next-line no-console
        console.error(error);
      } else {
        // eslint-disable-next-line no-console
        console.error(err);
      }
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e);
    }
  },
};

const getDefaultNetworkHelpers = () => DEFAULT_NETWORK_HELPERS;

export default getDefaultNetworkHelpers;
