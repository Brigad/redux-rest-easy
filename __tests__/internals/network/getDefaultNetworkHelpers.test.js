import getDefaultNetworkHelpers from '../../../src/internals/network/getDefaultNetworkHelpers';

const NETWORK_HELPERS = getDefaultNetworkHelpers();

describe('getDefaultNetworkHelpers', () => {
  test('only path', () => {
    expect(NETWORK_HELPERS).toMatchSnapshot();
  });
});

describe('requestGET', () => {
  test('only path', () => {
    expect(NETWORK_HELPERS.requestGET()).toMatchSnapshot();
  });
});

describe('requestPATCH', () => {
  test('only path', () => {
    expect(NETWORK_HELPERS.requestPATCH()).toMatchSnapshot();
  });
});

describe('requestPUT', () => {
  test('only path', () => {
    expect(NETWORK_HELPERS.requestPUT()).toMatchSnapshot();
  });
});

describe('requestPOST', () => {
  test('only path', () => {
    expect(NETWORK_HELPERS.requestPOST()).toMatchSnapshot();
  });
});

describe('requestDELETE', () => {
  test('only path', () => {
    expect(NETWORK_HELPERS.requestDELETE()).toMatchSnapshot();
  });
});

describe('handleStatusCode', () => {
  test('valid response', () => {
    const response = { status: 200 };
    const response2 = { status: 299 };

    expect(NETWORK_HELPERS.handleStatusCode(response)).toBe(response);
    expect(NETWORK_HELPERS.handleStatusCode(response2)).toBe(response2);
  });

  test('invalid response', () => {
    const response = { status: 300 };

    expect(() => NETWORK_HELPERS.handleStatusCode(response)).toThrow();
  });
});

describe('handleError', () => {
  // eslint-disable-next-line no-console
  const INITIAL_CONSOLE_ERROR = console.error;
  let consoleErrorMock;

  beforeEach(() => {
    consoleErrorMock = jest.fn();
    // eslint-disable-next-line no-console
    console.error = consoleErrorMock;
  });

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error = INITIAL_CONSOLE_ERROR;
  });

  test('wrongly formatted error', async () => {
    const error = 'oh no';

    await NETWORK_HELPERS.handleError(error);

    expect(consoleErrorMock).toHaveBeenCalledTimes(1);
  });
});
