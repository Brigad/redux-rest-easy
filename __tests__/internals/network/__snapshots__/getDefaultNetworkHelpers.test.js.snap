// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDefaultNetworkHelpers only path 1`] = `
Object {
  "getToken": [Function],
  "handleError": [Function],
  "handleStatusCode": [Function],
  "requestDELETE": [Function],
  "requestGET": [Function],
  "requestPATCH": [Function],
  "requestPOST": [Function],
  "requestPUT": [Function],
}
`;

exports[`requestDELETE only path 1`] = `
Object {
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer token",
  },
  "method": "DELETE",
}
`;

exports[`requestGET only path 1`] = `
Object {
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer token",
  },
  "method": "GET",
}
`;

exports[`requestPATCH only path 1`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer token",
    "Content-Type": "application/json",
  },
  "method": "PATCH",
}
`;

exports[`requestPOST only path 1`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer token",
    "Content-Type": "application/json",
  },
  "method": "POST",
}
`;

exports[`requestPUT only path 1`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Accept": "application/json",
    "Authorization": "Bearer token",
    "Content-Type": "application/json",
  },
  "method": "PUT",
}
`;
